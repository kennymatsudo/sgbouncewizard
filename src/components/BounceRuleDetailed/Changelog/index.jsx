import React from "react";
import {
  HeaderCell,
  Table,
  TableBody,
  TableCell,
  TableHeader,
  TableRow,
} from "@sendgrid/ui-components/table/table";
import Card from "@sendgrid/ui-components/card";
import { CSVLink } from "react-csv";
import moment from "moment";
import { Row } from "../../Row";
import { Column } from "../../Column";

import { WriteSelectors } from "../selectors";

function showChanges(changelog, rulesToShow, handleChangelogClicked) {
  return changelog
    .slice(0, rulesToShow)
    .map((change, index) => (
      <Changes
        key={change.created_at}
        index={index}
        change={change}
        handleChangelogClicked={handleChangelogClicked}
      />
    ));
}

const EmptyChangelog = () => (
  <Row>
    <Card
      icon="airplane"
      iconSize={40}
      title="No Previous Changes Found!"
      body="No changes have been previously made to this bounce rule."
      centered
    />
  </Row>
);

const Changelog = ({
  handleChangelogClicked,
  changelog,
  isChangelogEmpty,
  rulesToShow,
}) => (
  <div {...WriteSelectors.changelog}>
    <Row id="changelog-title">
      <Column width={1} offset={1}>
        <h2>Changelog</h2>
      </Column>
      <Column width={2} offset={11} className="changelog-csv sg-right">
        <CSVLink data={changelog} filename="changelog.csv" target="_blank">
          Export as CSV
        </CSVLink>
      </Column>
    </Row>
    {isChangelogEmpty && <EmptyChangelog />}
    {!isChangelogEmpty && (
      <Table>
        <TableHeader>
          <TableRow>
            <HeaderCell>Date</HeaderCell>
            <HeaderCell>User</HeaderCell>
            <HeaderCell>Commit Message</HeaderCell>
            <HeaderCell className="actions-align">Actions</HeaderCell>
          </TableRow>
        </TableHeader>
        <TableBody>
          {showChanges(changelog, rulesToShow, handleChangelogClicked)}
        </TableBody>
      </Table>
    )}
  </div>
);

const Changes = ({ change, handleChangelogClicked, index }) => {
  const { created_at: createdAt, user_id: userId, comment } = change;

  return (
    <TableRow>
      <TableCell>{moment.unix(createdAt).format("MM/DD/YYYY LTS")}</TableCell>
      <TableCell>{userId}</TableCell>
      <TableCell>{comment}</TableCell>
      <TableCell className="changelog-view-icon-cell">
        <i
          onClick={handleChangelogClicked}
          id="changeClicked"
          index={index}
          onKeyDown={handleChangelogClicked}
          className="sg-icon sg-icon-view changelog-view-icon"
          role="button"
          tabIndex={0}
        />
      </TableCell>
    </TableRow>
  );
};

export default Changelog;
