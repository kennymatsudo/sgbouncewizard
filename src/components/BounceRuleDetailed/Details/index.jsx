import React from "react";
import { TextInput } from "@sendgrid/ui-components/text-input";
import {
  Table,
  TableBody,
  TableCell,
  TableRow,
} from "@sendgrid/ui-components/table/table";
import { WriteSelectors } from "../selectors";

const DetailsContainer = ({ currentRule, handleEditClicked }) => {
  const {
    description,
    id,
    response_code: responseCode,
    enhanced_code: enhancedCode,
    bounce_action: bounceAction,
    regex,
    priority,
  } = currentRule;
  return (
    <div
      onClick={handleEditClicked}
      id="isEditClicked"
      onKeyDown={handleEditClicked}
      role="searchbox"
      tabIndex={0}
      className="detail-container card "
    >
      <div className="editable">
        <i className="sg-icon sg-icon-editor-design" />
      </div>
      <div className="description-info">
        <Table className="table-fixed">
          <TableBody>
            <TableRow>
              <TableCell className="description-cell">
                <strong>Description</strong>
              </TableCell>
              <TableCell> {description} </TableCell>
            </TableRow>
          </TableBody>
        </Table>
      </div>
      <div className="detail-info">
        <Table>
          <TableBody>
            <TableRow>
              <TableCell>
                <strong>Bounce ID</strong>
              </TableCell>
              <TableCell> {id} </TableCell>
            </TableRow>
            <TableRow>
              <TableCell>
                <strong>Response Code</strong>
              </TableCell>
              <TableCell> {responseCode} </TableCell>
            </TableRow>
            <TableRow>
              <TableCell>
                <strong>Enhanced Code</strong>
              </TableCell>
              <TableCell> {enhancedCode} </TableCell>
            </TableRow>
          </TableBody>
        </Table>
        <Table>
          <TableBody>
            <TableRow>
              <TableCell>
                <strong>Regex</strong>
              </TableCell>
              <TableCell> {regex} </TableCell>
            </TableRow>
            <TableRow>
              <TableCell>
                <strong>Priority</strong>
              </TableCell>
              <TableCell> {priority} </TableCell>
            </TableRow>
            <TableRow>
              <TableCell>
                <strong>Bounce Action</strong>
              </TableCell>
              <TableCell> {bounceAction} </TableCell>
            </TableRow>
          </TableBody>
        </Table>
      </div>
    </div>
  );
};

export const DetailsContainerEditable = ({
  onChangeRule,
  onChangeRuleInt,
  updatedRule,
}) => {
  const {
    description,
    id,
    response_code: responseCode,
    enhanced_code: enhancedCode,
    bounce_action: bounceAction,
    regex,
    priority,
  } = updatedRule;
  return (
    <div className="detail-container detail-container-editable card ">
      <div className="editable">
        <i className="sg-icon sg-icon-editor-design" />
      </div>
      <div className="description-info">
        <Table className="table-fixed">
          <TableBody>
            <TableRow>
              <TableCell className="description-cell">
                <strong>Description</strong>
              </TableCell>
              <TableCell>
                <TextInput
                  {...WriteSelectors.description}
                  onChange={onChangeRule}
                  id="description"
                  value={description}
                  type="text"
                />
              </TableCell>
            </TableRow>
          </TableBody>
        </Table>
      </div>
      <div className="detail-info">
        <Table>
          <TableBody>
            <TableRow>
              <TableCell>
                <strong>Bounce ID</strong>
              </TableCell>
              <TableCell>
                <div className="uneditable">
                  <strong>{id}</strong>
                </div>
              </TableCell>
            </TableRow>
            <TableRow>
              <TableCell>
                <strong>Response Code</strong>
              </TableCell>
              <TableCell>
                <TextInput
                  {...WriteSelectors.responseCode}
                  onChange={onChangeRuleInt}
                  id="response_code"
                  value={responseCode}
                  type="number"
                />
              </TableCell>
            </TableRow>
            <TableRow>
              <TableCell>
                <strong>Enhanced Code</strong>
              </TableCell>
              <TableCell>
                <TextInput
                  {...WriteSelectors.enhancedCode}
                  onChange={onChangeRule}
                  id="enhanced_code"
                  value={enhancedCode}
                  type="text"
                />
              </TableCell>
            </TableRow>
          </TableBody>
        </Table>
        <Table>
          <TableBody>
            <TableRow>
              <TableCell>
                <strong>Regex</strong>
              </TableCell>
              <TableCell>
                <TextInput
                  {...WriteSelectors.regex}
                  onChange={onChangeRule}
                  id="regex"
                  value={regex}
                  type="text"
                />
              </TableCell>
            </TableRow>
            <TableRow>
              <TableCell>
                <strong>Priority</strong>
              </TableCell>
              <TableCell>
                <TextInput
                  {...WriteSelectors.priority}
                  onChange={onChangeRuleInt}
                  id="priority"
                  value={priority}
                  type="number"
                />
              </TableCell>
            </TableRow>
            <TableRow>
              <TableCell>
                <strong>Bounce Action</strong>
              </TableCell>
              <TableCell>
                <TextInput
                  {...WriteSelectors.bounceAction}
                  onChange={onChangeRule}
                  id="bounce_action"
                  value={bounceAction}
                  type="text"
                />
              </TableCell>
            </TableRow>
          </TableBody>
        </Table>
      </div>
    </div>
  );
};

export default DetailsContainer;
