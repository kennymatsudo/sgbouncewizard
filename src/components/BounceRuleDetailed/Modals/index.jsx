import React from "react";
import CenterModal from "@sendgrid/ui-components/center-modal";
import Button from "@sendgrid/ui-components/button";
import {
  Table,
  TableBody,
  TableCell,
  TableRow,
} from "@sendgrid/ui-components/table/table";
import { Column } from "../../Column";
import { Row } from "../../Row";

const ConfirmModalBody = ({
  isConfirmOpen,
  isCancelConfirmOpen,
  handleModalClose,
  handleModalConfirm,
  modalType,
}) => (
  <div>
    <Row>
      <Column>
        {(isConfirmOpen && (
          <div>
            <h2>Are you sure you&apos;d like to commit these changes?</h2>
            <p>
              {
                "Doing so will effect how current email will be handled via this\
                bounce rule. This action will go into effect immediately."
              }
            </p>
          </div>
        )) ||
          (isCancelConfirmOpen && (
            <div>
              <h2>Are you sure you want to discard these changes?</h2>
              <p>Doing so will remove applied changes.</p>
            </div>
          ))}
      </Column>
    </Row>
    <Row>
      <Column width={1} offset={10}>
        <Button
          className="sg-button"
          onClick={handleModalClose}
          id={modalType}
          type="secondary"
        >
          {"Close"}
        </Button>
      </Column>
      <Column width={1} offset={11}>
        <Button
          className="sg-button"
          id={modalType}
          onClick={handleModalConfirm}
          type="primary"
        >
          {"Confirm"}
        </Button>
      </Column>
    </Row>
  </div>
);

const ChangeModalBody = ({ currentRule, handleModalClose }) => (
  <div className="changelog-modal">
    <Row>
      <Column width={6} offset={1}>
        <h1>Current</h1>
        <ChangeTable currentRule={currentRule} />
      </Column>
      <Column width={6} offset={7}>
        <h1>Previous</h1>
        <ChangeTable currentRule={currentRule} />
      </Column>
    </Row>
    <Row>
      <Column className="changelog-modal-button sg-right" width={1} offset={12}>
        <Button id="changeModal" onClick={handleModalClose}>
          Close
        </Button>
      </Column>
    </Row>
  </div>
);

const ChangeTable = ({ currentRule }) => {
  const {
    description,
    id,
    response_code: responseCode,
    enhanced_code: enhancedCode,
    bounce_action: bounceAction,
    regex,
    priority,
  } = currentRule;
  return (
    <Table className="change-table">
      <TableBody>
        <TableRow>
          <TableCell>
            <strong>ID</strong>
          </TableCell>
          <TableCell>{id}</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>
            <strong>Bounce Action</strong>
          </TableCell>
          <TableCell>{bounceAction}</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>
            <strong>Response Code</strong>
          </TableCell>
          <TableCell>{responseCode}</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>
            <strong>Enhanced Code</strong>
          </TableCell>
          <TableCell>{enhancedCode}</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>
            <strong>RegEx</strong>
          </TableCell>
          <TableCell>{regex}</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>
            <strong>Priority</strong>
          </TableCell>
          <TableCell>{priority}</TableCell>
        </TableRow>
        <TableRow>
          <TableCell>
            <strong>Description</strong>
          </TableCell>
          <TableCell>{description}</TableCell>
        </TableRow>
      </TableBody>
    </Table>
  );
};

const ModalDisplay = ({
  isChangeModalOpen,
  isConfirmOpen,
  isCancelConfirmOpen,
  currentRule,
  handleModalClose,
  handleModalConfirm,
}) => (
  <CenterModal
    large={isChangeModalOpen}
    open={isChangeModalOpen || isConfirmOpen || isCancelConfirmOpen}
    renderBody={
      (isChangeModalOpen && (
        <ChangeModalBody
          handleModalClose={handleModalClose}
          currentRule={currentRule}
        />
      )) ||
      (isConfirmOpen && (
        <ConfirmModalBody
          isConfirmOpen={isConfirmOpen}
          handleModalClose={handleModalClose}
          handleModalConfirm={handleModalConfirm}
          modalType="saveModal"
        />
      )) ||
      (isCancelConfirmOpen && (
        <ConfirmModalBody
          isCancelConfirmOpen={isCancelConfirmOpen}
          handleModalClose={handleModalClose}
          currentRule={currentRule}
          handleModalConfirm={handleModalConfirm}
          modalType="cancelModal"
        />
      ))
    }
  />
);

export default ModalDisplay;
