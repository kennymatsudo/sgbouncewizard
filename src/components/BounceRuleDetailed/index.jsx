import React from "react";
import Breadcrumb from "@sendgrid/ui-components/breadcrumb";
import CenterModal from "@sendgrid/ui-components/center-modal";
import Button from "@sendgrid/ui-components/button";
import { TextInput } from "@sendgrid/ui-components/text-input";
import { Link } from "react-router-dom";
import {
  HeaderCell,
  Table,
  TableBody,
  TableCell,
  TableHeader,
  TableRow,
} from "@sendgrid/ui-components/table/table";
import { Row } from "../Row";
import { Column } from "../Column";
import Header from "../Header";
import Pagination from "../Pagination";
import "./index.scss";
import mockChangelog from "../../mocks/index";

const DetailContainer = ({ currentRule, handleButtonClicked }) => (
  <div
    onClick={() => handleButtonClicked("editClicked")}
    onKeyDown={() => handleButtonClicked("editClicked")}
    role="searchbox"
    tabIndex={0}
    className="detail-container card "
  >
    <div className="editable">
      <i className="sg-icon sg-icon-editor-design" />
    </div>
    <div className="description-info">
      <Table className="table-fixed">
        <TableBody>
          <TableRow>
            <TableCell className="description-cell">
              <strong>Description</strong>
            </TableCell>
            <TableCell> {currentRule.description} </TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </div>
    <div className="detail-info">
      <Table>
        <TableBody>
          <TableRow>
            <TableCell>
              <strong>Bounce ID</strong>
            </TableCell>
            <TableCell> {currentRule.id} </TableCell>
          </TableRow>
          <TableRow>
            <TableCell>
              <strong>Response Code</strong>
            </TableCell>
            <TableCell> {currentRule.response_code} </TableCell>
          </TableRow>
          <TableRow>
            <TableCell>
              <strong>Enhanced Code</strong>
            </TableCell>
            <TableCell> {currentRule.enhanced_code} </TableCell>
          </TableRow>
        </TableBody>
      </Table>
      <Table>
        <TableBody>
          <TableRow>
            <TableCell>
              <strong>Regex</strong>
            </TableCell>
            <TableCell> {currentRule.regex} </TableCell>
          </TableRow>
          <TableRow>
            <TableCell>
              <strong>Priority</strong>
            </TableCell>
            <TableCell> {currentRule.priority} </TableCell>
          </TableRow>
          <TableRow>
            <TableCell>
              <strong>Bounce Action</strong>
            </TableCell>
            <TableCell> {currentRule.bounce_action} </TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </div>
  </div>
);

const DetailContainerEditable = ({ currentRule, onChangeRule }) => (
  <div className="detail-container detail-container-editable card ">
    <div className="editable">
      <i className="sg-icon sg-icon-editor-design" />
    </div>
    <div className="description-info">
      <Table className="table-fixed">
        <TableBody>
          <TableRow>
            <TableCell className="description-cell">
              <strong>Description</strong>
            </TableCell>
            <TableCell>
              <TextInput
                onChange={e => onChangeRule(e, "description")}
                value={currentRule.description}
                type="text"
                label={currentRule.description}
              />
            </TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </div>
    <div className="detail-info">
      <Table>
        <TableBody>
          <TableRow>
            <TableCell>
              <strong>Bounce ID</strong>
            </TableCell>
            <TableCell>
              <TextInput
                onChange={e => onChangeRule(e, "id")}
                value={currentRule.id}
                type="text"
                label={currentRule.id}
              />
            </TableCell>
          </TableRow>
          <TableRow>
            <TableCell>
              <strong>Response Code</strong>
            </TableCell>
            <TableCell>
              <TextInput
                onChange={e => onChangeRule(e, "response_code")}
                value={currentRule.response_code}
                type="text"
                label={currentRule.response_code}
              />
            </TableCell>
          </TableRow>
          <TableRow>
            <TableCell>
              <strong>Enhanced Code</strong>
            </TableCell>
            <TableCell>
              <TextInput
                onChange={e => onChangeRule(e, "enhanced_code")}
                value={currentRule.enhanced_code}
                type="text"
                label={currentRule.enhanced_code}
              />
            </TableCell>
          </TableRow>
        </TableBody>
      </Table>
      <Table>
        <TableBody>
          <TableRow>
            <TableCell>
              <strong>Regex</strong>
            </TableCell>
            <TableCell>
              <TextInput
                onChange={e => onChangeRule(e, "regex")}
                value={currentRule.regex}
                type="text"
                label={currentRule.regex}
              />
            </TableCell>
          </TableRow>
          <TableRow>
            <TableCell>
              <strong>Priority</strong>
            </TableCell>
            <TableCell>
              <TextInput
                onChange={e => onChangeRule(e, "priority")}
                value={currentRule.priority}
                type="text"
                label={currentRule.priority}
              />
            </TableCell>
          </TableRow>
          <TableRow>
            <TableCell>
              <strong>Bounce Action</strong>
            </TableCell>
            <TableCell>
              <TextInput
                onChange={e => onChangeRule(e, "bounce_action")}
                value={currentRule.bounce_action}
                type="text"
                label={currentRule.bounce_action}
              />
            </TableCell>
          </TableRow>
        </TableBody>
      </Table>
    </div>
  </div>
);

const Changelog = ({ handleButtonClicked }) => (
  <div>
    <h2>Changelog</h2>
    <Table>
      <TableHeader>
        <TableRow>
          <HeaderCell>Date</HeaderCell>
          <HeaderCell>User</HeaderCell>
          <HeaderCell>Commit Message</HeaderCell>
          <HeaderCell className="actions-align">Actions</HeaderCell>
        </TableRow>
      </TableHeader>
      <TableBody>
        {mockChangelog.map(change => (
          <ChangelogMin
            key={change.id}
            change={change}
            handleButtonClicked={handleButtonClicked}
          />
        ))}
      </TableBody>
    </Table>
  </div>
);

const ChangelogMin = ({ change, handleButtonClicked }) => (
  <TableRow>
    <TableCell>{change.date}</TableCell>
    <TableCell>{change.user}</TableCell>
    <TableCell>{change.message}</TableCell>
    <TableCell className="changelog-view-icon-cell">
      <i
        onClick={() => handleButtonClicked("changeClicked")}
        onKeyDown={() => handleButtonClicked("changeClicked")}
        className="sg-icon sg-icon-view changelog-view-icon"
        role="button"
        tabIndex={0}
      />
    </TableCell>
  </TableRow>
);

const ConfirmModalBody = ({
  isConfirmOpen,
  isCancelConfirmOpen,
  handleModalClose,
  handleModalConfirm,
}) => (
  <div>
    <Row>
      <Column>
        {(isConfirmOpen && (
          <div>
            <h2>Are you sure you&apos;d like to commit these changes?</h2>
            <p>
              {
                "Doing so will effect how current email will be handled via this\
              bounce rule. This action will go into effect immediately."
              }
            </p>
          </div>
        )) ||
          (isCancelConfirmOpen && (
            <div>
              <h2>Are you sure you want to discard these changes?</h2>
              <p>Doing so will remove applied changes.</p>
            </div>
          ))}
      </Column>
    </Row>
    <Row>
      <Column width={1} offset={10}>
        <Button
          className="sg-button"
          onClick={handleModalClose}
          type="secondary"
        >
          {"Close"}
        </Button>
      </Column>
      <Column width={1} offset={11}>
        <Button
          className="sg-button"
          onClick={handleModalConfirm}
          type="primary"
        >
          {"Submit"}
        </Button>
      </Column>
    </Row>
  </div>
);

const ChangeModalBody = ({ currentRule, handleModalClose }) => (
  <div className="changelog-modal">
    <Row>
      <Column width={6} offset={1}>
        <h1>Current</h1>
        <ChangeTable currentRule={currentRule} />
      </Column>
      <Column width={6} offset={7}>
        <h1>Previous</h1>
        <ChangeTable currentRule={currentRule} />
      </Column>
    </Row>
    <Row>
      <Column className="changelog-modal-button sg-right" width={1} offset={12}>
        <Button onClick={handleModalClose}>Close</Button>
      </Column>
    </Row>
  </div>
);

const ChangeTable = ({ currentRule }) => (
  <Table className="change-table">
    <TableBody>
      <TableRow>
        <TableCell>
          <strong>ID</strong>
        </TableCell>
        <TableCell>{currentRule.id}</TableCell>
      </TableRow>
      <TableRow>
        <TableCell>
          <strong>Bounce Action</strong>
        </TableCell>
        <TableCell>{currentRule.bounce_action}</TableCell>
      </TableRow>
      <TableRow>
        <TableCell>
          <strong>Response Code</strong>
        </TableCell>
        <TableCell>{currentRule.response_code}</TableCell>
      </TableRow>
      <TableRow>
        <TableCell>
          <strong>Enhanced Code</strong>
        </TableCell>
        <TableCell>{currentRule.enhanced_code}</TableCell>
      </TableRow>
      <TableRow>
        <TableCell>
          <strong>RegEx</strong>
        </TableCell>
        <TableCell>{currentRule.regex}</TableCell>
      </TableRow>
      <TableRow>
        <TableCell>
          <strong>Priority</strong>
        </TableCell>
        <TableCell>{currentRule.priority}</TableCell>
      </TableRow>
      <TableRow>
        <TableCell>
          <strong>Description</strong>
        </TableCell>
        <TableCell>{currentRule.description}</TableCell>
      </TableRow>
    </TableBody>
  </Table>
);

const BounceRuleDetailed = ({
  currentRule,
  newRule,
  isEditClicked,
  isChangeModalOpen,
  isConfirmOpen,
  isCancelConfirmOpen,
  handleModalConfirm,
  handleModalClose,
  onChangeRule,
  handleButtonClicked,
}) => (
  <div className="detailed-page-container">
    <Header />
    <Row>
      <Column width={6} offset={2}>
        <Breadcrumb>
          <Link to="/bounce_rules">Bounce Rules</Link>
          <Link to={`/bounce_rules/${currentRule.id}`}>{currentRule.id}</Link>
        </Breadcrumb>
      </Column>
    </Row>
    <Row>
      <Column width={6} offset={2}>
        <h1>
          {"Bounce Rule"} {currentRule.id}
        </h1>
      </Column>
      {isEditClicked ? (
        <Column className=" sg-right" width={4} offset={8}>
          <Button
            onClick={() => handleButtonClicked("cancelClicked")}
            onKeyDown={() => handleButtonClicked("cancelClicked")}
            data-test="cancel-button"
            className="header-button cancel-button"
            type="secondary"
          >
            {"Cancel"}
          </Button>
          <Button
            onClick={() => handleButtonClicked("saveClicked")}
            onKeyDown={() => handleButtonClicked("saveClicked")}
            data-test="save-button"
            className="header-button save-button"
            type="primary"
          >
            {"Save"}
          </Button>
        </Column>
      ) : (
        <Column
          className="details-button-column sg-right"
          width={1}
          offset={11}
        >
          <span>
            <Button
              onClick={() => handleButtonClicked("editClicked")}
              onKeyDown={() => handleButtonClicked("editClicked")}
              data-test="edit-button"
              className="header-button edit-button"
              type="primary"
            >
              {"Edit Rule"}
            </Button>
          </span>
        </Column>
      )}
      {/* </Column> */}
    </Row>
    <Row>
      <Column width={10} offset={2}>
        <div>
          {isEditClicked ? (
            <DetailContainerEditable
              data-test="detailed-container-editable"
              currentRule={currentRule}
              onChangeRule={onChangeRule}
              newRule={newRule}
            />
          ) : (
            <DetailContainer
              data-test="detailed-container"
              currentRule={currentRule}
              handleButtonClicked={handleButtonClicked}
            />
          )}
        </div>
      </Column>
    </Row>
    <Row className="changelog-container">
      <Column width={10} offset={2}>
        <Changelog
          data-test="changelog-container"
          handleButtonClicked={handleButtonClicked}
        />
      </Column>
    </Row>
    <Row>
      <Column width={4} offset={5}>
        <Pagination
          prevPageIndex={1}
          nextPageIndex={10}
          pageIndex={1}
          pageInterval={1}
          numRules={3}
          updatePageIndex={() => {}}
        />
      </Column>
    </Row>

    <ModalDisplay
      data-test="modal"
      isChangeModalOpen={isChangeModalOpen}
      isConfirmOpen={isConfirmOpen}
      isCancelConfirmOpen={isCancelConfirmOpen}
      currentRule={currentRule}
      handleModalClose={handleModalClose}
      handleModalConfirm={handleModalConfirm}
    />
  </div>
);

const ModalDisplay = ({
  isChangeModalOpen,
  isConfirmOpen,
  isCancelConfirmOpen,
  currentRule,
  handleModalClose,
  handleModalConfirm,
}) => (
  <CenterModal
    large={isChangeModalOpen}
    open={isChangeModalOpen || isConfirmOpen || isCancelConfirmOpen}
    renderBody={
      (isChangeModalOpen && (
        <ChangeModalBody
          handleModalClose={() => handleModalClose("changeModal")}
          currentRule={currentRule}
        />
      )) ||
      (isConfirmOpen && (
        <ConfirmModalBody
          isConfirmOpen={isConfirmOpen}
          handleModalClose={() => handleModalClose("saveModal")}
          handleModalConfirm={() => handleModalConfirm("saveModal")}
        />
      )) ||
      (isCancelConfirmOpen && (
        <ConfirmModalBody
          isCancelConfirmOpen={isCancelConfirmOpen}
          handleModalClose={() => handleModalClose("cancelModal")}
          currentRule={currentRule}
          handleModalConfirm={() => handleModalConfirm("cancelModal")}
        />
      ))
    }
  />
);

export default BounceRuleDetailed;
