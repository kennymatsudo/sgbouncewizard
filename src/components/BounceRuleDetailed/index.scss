.header-button {
  &.back-button {
    margin-right: 10px;
  }
}

.detail-container {
  display: flex;
  justify-content: center;
  align-items: center;
  height: 220px;
  padding: 5px;
  background: #f4f6f7;

  td {
    padding: 15px 10px;

    &.description-cell {
      width: 25%;
    }
  }

  th,
  td {
    word-wrap: break-word;
  }
}

.table-fixed {
  table-layout: fixed;
}

.description-info {
  width: 100%;
  .table-wrap {
    margin-bottom: 0;
  }
}

.detail-info {
  display: flex;
  width: 100%;
}

.changelog-container {
  margin-top: 2vh;

  th.actions-align {
    padding-right: 0;
    text-align: right;
  }
}

.changelog-view-icon-cell {
  text-align: right;
}

.changelog-view-icon {
  color: #7f90a0;
  &:hover {
    cursor: pointer;
  }
}

.changelog-modal {
  padding: 5vh;

  h1 {
    margin-bottom: 0px;
  }
  .change-table {
    > td {
      padding: 15px;
    }
  }
}
