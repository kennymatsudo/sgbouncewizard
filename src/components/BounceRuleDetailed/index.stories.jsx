import React from "react";
import { storiesOf } from "@storybook/react";
import { action } from "@storybook/addon-actions";
import StoryRouter from "storybook-react-router";
import "./index.scss";
import DetailsContainer, { DetailsContainerEditable } from "./Details";
import Changelog from "./Changelog";
import ChangeModal from "./Modals/ChangeModal";
import ConfirmationModal from "./Modals/ConfirmationModal";
import CancelConfirmationModal from "./Modals/CancelConfirmationModal";

const sampleRule = {
  id: 504,
  response_code: 550,
  enhanced_code: "",
  regex: "no MX record for domain",
  priority: 0,
  description:
    "mainly liberty domain block seeing ~50% of addresses engaging SG wide",
  bounce_action: "no_action",
  created_by: 99,
};

const sampleChangelog = {
  id: 300,
  response_code: 552,
  enhanced_code: "4.2.2",
  regex: "User has full mailbox",
  priority: 0,
  description:
    "message will not succeed when retried, but address is likely valid",
  bounce_action: "no_action",
  user_id: 1,
  created_at: 1542594709,
  comment: "Testing changelog 1",
};

class BounceRuleDetailsContainer extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      currentRule: sampleRule,
      updatedRule: sampleRule,
      isEditClicked: false,
      changelog: [sampleChangelog],
      ...this.props,
    };

    this.onChangeRule = this.onChangeRule.bind(this);
  }

  onChangeRule(e) {
    const { id, value } = e.currentTarget;
    const { updatedRule } = this.state;
    this.setState({
      updatedRule: { ...updatedRule, [id]: value },
    });
  }

  render() {
    const {
      render,
      currentRule,
      updatedRule,
      isEditClicked,
      changelog,
    } = this.state;
    return (
      <div>
        {render(
          this.onChangeRule,
          currentRule,
          updatedRule,
          isEditClicked,
          changelog
        )}
      </div>
    );
  }
}

storiesOf("Bounce Rule Details", module)
  .addDecorator(StoryRouter())
  .add("Default", () => <DetailsContainer currentRule={sampleRule} />)
  .add("Editable", () => (
    <BounceRuleDetailsContainer
      render={(
        onChangeRule,
        currentRule,
        updatedRule,
        isEditClicked,
        changelog
      ) => (
        <DetailsContainerEditable
          currentRule={currentRule}
          updatedRule={updatedRule}
          onChangeRule={onChangeRule}
          onChangeRuleInt={onChangeRule}
          changelog={changelog}
        />
      )}
    />
  ));

storiesOf("Bounce Rule Changelog", module)
  .add("Default", () => (
    <BounceRuleDetailsContainer
      render={(
        onChangeRule,
        currentRule,
        updatedRule,
        isEditClicked,
        changelog
      ) => <Changelog changelog={changelog} />}
    />
  ))
  .add("Empty", () => <Changelog changelog={[]} isChangelogEmpty />);

storiesOf("Modals", module)
  .add("Change Log Modal", () => (
    <ChangeModal
      currentRule={sampleRule}
      handleModalClose={action("close modal")}
      handleModalConfirm={action("open modal")}
    />
  ))
  .add("Confirmation Modal", () => (
    <ConfirmationModal
      currentRule={sampleRule}
      handleModalClose={action("close modal")}
      handleModalConfirm={action("open modal")}
    />
  ))
  .add("Cancellation Modal", () => (
    <CancelConfirmationModal
      currentRule={sampleRule}
      handleModalClose={action("close modal")}
      handleModalConfirm={action("open modal")}
    />
  ));
