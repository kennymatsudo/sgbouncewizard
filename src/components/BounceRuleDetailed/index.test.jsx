import React from "react";
import { shallow } from "enzyme";
import EnzymeToJson from "enzyme-to-json";
import BounceRuleDetailed from ".";

const sampleRule = {
  id: 504,
  response_code: "550",
  enhanced_code: "",
  regex: "no MX record for domain",
  priority: 0,
  description:
    "bWFpbmx5IGxpYmVydHkgZG9tYWluIGJsb2NrIHNlZWluZyB+NTAlIG9mIGFkZHJlc3NlcyBlbmdhZ2luZyBTRyB3aWRl",
  bounce_action: "no_action",
};

const DetailedObj = (
  <BounceRuleDetailed
    currentRule={sampleRule}
    isEditClicked={false}
    isChangeModalOpen={false}
    isCancelConfirmOpen={false}
    isConfirmOpen={false}
    handleModalClose={() => {}}
    handleButtonClicked={() => {}}
    handleModalConfirm={() => {}}
    onChangeRule={() => {}}
  />
);

describe("Bounce Rule Detailed", () => {
  let wrapper;
  beforeEach(() => {
    wrapper = shallow(DetailedObj);
  });

  it("should render correctly", () => {
    expect(EnzymeToJson(DetailedObj)).toMatchSnapshot();
  });

  it("should render details table", () => {
    expect(wrapper.find('[data-test="detailed-container"]')).toHaveLength(1);
  });

  it("should render change log table", () => {
    expect(wrapper.find('[data-test="changelog-container"]')).toHaveLength(1);
  });

  it("should show cancel/save button when edit is clicked", () => {
    expect(wrapper.find('[data-test="edit-button"]')).toHaveLength(1);
    wrapper.setProps({
      isEditClicked: true,
    });
    expect(wrapper.find('[data-test="cancel-button"]')).toHaveLength(1);
    expect(wrapper.find('[data-test="save-button"]')).toHaveLength(1);
  });

  it("should be editable when edit is clicked", () => {
    wrapper.setProps({
      isEditClicked: true,
    });
    expect(
      wrapper.find('[data-test="detailed-container-editable"]')
    ).toHaveLength(1);
  });

  it("should display cancel confirm modal when cancel is clicked", () => {
    wrapper.setProps({
      isCancelConfirmOpen: true,
    });
    expect(wrapper.find('[data-test="modal"]')).toHaveLength(1);
  });

  it("should display save confirm modal when save is clicked", () => {
    wrapper.setProps({
      isConfirmOpen: true,
    });
    expect(wrapper.find('[data-test="modal"]')).toHaveLength(1);
  });

  it("should display change modal when rule is clicked", () => {
    wrapper.setProps({
      isChangeModalOpen: true,
    });
    wrapper.update();
    expect(wrapper.find('[data-test="modal"]')).toHaveLength(1);
  });
});
