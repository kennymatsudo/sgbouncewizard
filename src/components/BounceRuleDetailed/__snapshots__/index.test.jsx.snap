// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bounce Rule Detailed should render correctly 1`] = `undefined`;
